tests/test_mask.py

```diff
@@ -0,0 +1,233 @@
+"""Tests for augur mask
+
+NOTE: Several functions are monkeypatched in these tests. If you change the arguments
+for any function in mask.py, check that it is correctly updated in this file.
+"""
+import argparse
+import os
+import pytest
+
+from Bio import SeqIO
+from Bio.Seq import Seq
+from Bio.SeqRecord import SeqRecord
+
+from augur import mask
+
+# Test inputs for the commands. Writing these here so tests are self-contained.
+@pytest.fixture
+def sequences():
+    return {
+        "SEQ1": SeqRecord(Seq("ATGC-ATGC-ATGC"), id="SEQ1"),
+        "SEQ2": SeqRecord(Seq("ATATATATATATATAT"), id="SEQ2")
+    }
+
+@pytest.fixture
+def fasta_file(tmpdir, sequences):
+    fasta_file = str(tmpdir / "test.fasta")
+    with open(fasta_file, "w") as fh:
+        SeqIO.write(sequences.values(), fh, "fasta")
+    return fasta_file
+
+TEST_VCF="""\
+##fileformat=VCFv4.2
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT
+SEQ	1	.	G	A	.		.
+SEQ	2	.	G	A	.		.
+SEQ	3	.	C	T	.		.
+SEQ	5	.	C	T	.		.
+SEQ	8	.	A	C	.		.
+"""
+
+@pytest.fixture
+def vcf_file(tmpdir):
+    vcf_file = str(tmpdir / "test.vcf")
+    with open(vcf_file, "w") as fh:
+        fh.write(TEST_VCF)
+    return vcf_file
+
+TEST_BED_SEQUENCE = [1,2,4,6,7,8,9,10]
+# IF YOU UPDATE ONE OF THESE, UPDATE THE OTHER.
+TEST_BED="""\
+Chrom	ChromStart	ChromEnd	locus tag	Comment	
+SEQ1	1	2	IG18_Rv0018c-Rv0019c	
+SEQ1	4	4	IG18_Rv0018c-Rv0019c	
+SEQ1	6	8	IG18_Rv0018c-Rv0019c	
+SEQ1	7	10	IG18_Rv0018c-Rv0019c	
+"""
+
+@pytest.fixture
+def bed_file(tmpdir):
+    bed_file = str(tmpdir / "exclude.bed")
+    with open(bed_file, "w") as fh:
+        fh.write(TEST_BED)
+    return bed_file
+
+@pytest.fixture
+def mp_context(monkeypatch):
+    #Have found explicit monkeypatch context-ing prevents stupid bugs
```

Comment:
I haven't seen this pattern before, so I'm interested to learn more about what kinds of bugs this avoids. Maybe better for an offline discussion unless there are simple examples you could mention here though.

Reply:
This goes back a couple years, but the basic pattern we saw was when a test failed, it wouldn't make it to the `monkeypatch.undo()` line, and subsequent tests were running with the monkeypatches still in place. `monkeypatch.context()` is basically a `try: monkeypatch(); finally: monkeypatch.undo()` block that ensures the patches get unset regardless of how the test exits.